### ISSUE TITLE
Universe polymorphism weakens [congruence]

### ISSUE THREAD

**Issue by coqbot** [coq/coq#5481]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#5481
From: @andersk
Reported version: 8.6
CC:   @herbelin

**Comment by PierreCorbineau** [comment]
Si this still active? IMO, "congruence" should force unification of universes if it means success.

**Comment by Alizter** [comment]
Still an issue in 8.15+alpha.

**Comment by JasonGross** [comment]
I agree with @PierreCorbineau, `congruence` should be willing to unify universes if otherwise it would fail.  Or this behavior can be controlled by a flag, but I'd like to at least see an option to fix it.  (I'm running into this issue trying to make MetaCoq more universe-polymorphic.)

**Comment by JasonGross** [comment]
This also impacts `f_equal` as per https://github.com/coq/coq/issues/9979#issuecomment-489443198

**Comment by herbelin** [comment]
Not fully clear at which level to act.

Unifying all subterms that differ only in universes seem overkill. I guess that we'd like to identify only those who are involved in the proof found by congruence?

So, what about dropping the universes when building the classes but reunifying the terms intended to be equal at the time of building the proof (in ccproof.ml, prefl, ptrans, ...).

Is there already around a hash function that drops universes?

**Comment by herbelin** [comment]
I tried a fix at #18106.

**Comment by herbelin** [comment]
There is an incident question...

The following succeeds:
```coq
Goal Prop -> Prop.
congruence.
```
The following fails (even w/o polymorphism):
```coq
Goal Type -> Type.
Fail congruence.
```
Do we make it succeed (by unifying the type levels)?
